### PR TITLE
Fixes #29963 - Manage services without foreman-maintain

### PR DESCRIPTION
**Pull request by ekohl**
Currently foreman-maintain is called to stop and start services in the EL7 upgrade. However, foreman-maintain is unavailable on Debian. It's also possibly a circular problem because foreman-maintain depends on the installer for certain actions so it is possible to end up in an infinite loop.

Successor to https://github.com/theforeman/foreman-installer/pull/461

Currently not entirely tested. The list of service names is also likely to be incomplete.

### CODE REVIEW

**Comment by ekohl** [comment]
After this is merged, we can also drop https://github.com/theforeman/foreman-packaging/blob/9a6c36d20882630001d61d59f81f1e3197f603e5/packages/foreman/foreman-installer/foreman-installer.spec#L26

File: hooks/boot/04-services.rb
Comment (ekohl) [inline hooks/boot/04-services.rb:4]: I already know qpid and friends is currently missing. Question is, do we wait for qpid to disappear or also list it here?

**Comment by ehelms** [comment]
And add it here https://github.com/theforeman/foreman-packaging/blob/9a6c36d20882630001d61d59f81f1e3197f603e5/packages/foreman/foreman-installer/foreman-installer.spec#L50

I get the theoretical concern around circular problems (though in a couple years we have yet to face any). And I get the issues with foreman-maintain not being available on Debian (though outside of packaging it I don't see why it couldn't be present on Debian). 

My concern is that this duplicates logic and it creates division between Foreman and Katello scenarios where our goal has been to bring congruence.

File: hooks/boot/04-services.rb
Comment (ehelms) [inline hooks/boot/04-services.rb:12]: If we are considering all services that are stopped as of today:

 * tomcat
 * puppet
 * puppetserver
 * smart_proxy_dynflow_core

Things going away soon:

 * qpidd
 * qdrouterd
 * squid

**Comment by ekohl** [comment]
> I get the theoretical concern around circular problems (though in a couple years we have yet to face any).

Likely because until recently the installer didn't depend on foreman-maintain.

> My concern is that this duplicates logic and it creates division between Foreman and Katello scenarios where our goal has been to bring congruence.

On the other hand, I'd argue this is much better logic since it's much more focused and it utilizes the parallelization that systemd offers.

Also note that we don't care about starting all services, just the individual ones we need for an action. That means we can get away with globbing.

We can also improve things. For example, resetting the databases shouldn't require us to stop Apache. If the foreman service is unavailable, we can render a proper HTTP 502 page. Resetting the databases also doesn't require us to stop puppet and puppetserver. If we're going that route, there really is no point in using foreman-maintain.

**Comment by ehelms** [comment]
> Likely because until recently the installer didn't depend on foreman-maintain.
> 
> >

Just adding data to the discussion, been relying on it since over 1.5+ years, dating back to the great merging of the installers even (https://github.com/theforeman/foreman-installer/commit/e8ce73a008d882aaf14ce31198aeac5624fd28a3#diff-ceb7d70cfab1a466aef1539a90401046)

File: hooks/boot/04-services.rb
Comment (ehelms) [reply hooks/boot/04-services.rb:12]: Can you add tomcat to the list?

File: hooks/boot/04-services.rb
Comment (ehelms) [inline hooks/boot/04-services.rb:21]: Are you expecting Kafo to catch this and deal with it? Otherwise I think we can to use `fail`

File: hooks/boot/04-services.rb
Comment (ehelms) [inline hooks/boot/04-services.rb:25]: Whats the reason behind the combined stdout and stderr here? (vs just putting stderr out)

**Comment by ehelms** [comment]
Can you fix the tests up.

**Comment by ekohl** [comment]
> Can you fix the tests up.

That's IMHO just Rubocop being stupid.

File: hooks/boot/04-services.rb
Comment (ekohl) [reply hooks/boot/04-services.rb:21]: Passing an empty array is incorrect API use and the goal of this is to slap the developer. Users should not see this. The alternative is to just return.

File: hooks/boot/04-services.rb
Comment (ekohl) [reply hooks/boot/04-services.rb:25]: I wasn't quite sure about the output. just grabbing the entire output allows to debug was my reasoning. Didn't think it through that much.

**Comment by ehelms** [comment]
Rules are rules it blocks this change from going in :)

File: hooks/boot/04-services.rb
Comment (ehelms) [reply hooks/boot/04-services.rb:25]: Suppose its a starting place.

File: hooks/boot/04-services.rb
Comment (ehelms) [reply hooks/boot/04-services.rb:21]: This was mostly around the nuance of raise vs fail in Ruby -- https://gist.github.com/Integralist/8058ac54088408d9aadc

Comment (ekohl) [reply hooks/boot/04-services.rb:21]: TIL. Thanks!

**Comment by ekohl** [comment]
So I've updated it and the tests pass. I'll see about testing this on a real system.

**Comment by ehelms** [comment]
Is this ready to be taken out of Draft?

**Comment by ekohl** [comment]
Code wise I think this is ready but I haven't tested this myself on an actual system.

**Comment by ekohl** [comment]
Testing on a real system showed that I had 2 issues:
* Empty output was not dealt with
* The incorrect variable was used (services instead of running)

Will update and expand the tests to cover this.

**Comment by ekohl** [comment]
Now tested on a real system, both with httpd.service running and not running.

File: hooks/boot/04-services.rb
Comment (ehelms) [inline hooks/boot/04-services.rb:20]: Can this default to ALL_POSSIBLE_SERVICES similar to stop_srevices?

Comment (ehelms) [reply hooks/boot/04-services.rb:20]: I see this doesn't handle well from systemd:

```
Failed to start services: Failed to start apache.service: Unit not found.
Failed to start smart_proxy_dynflow_core.service: Unit not found.
[ERROR 2020-07-01T14:57:11 verbose] Failed to start services: Failed to start apache.service: Unit not found.
Failed to start smart_proxy_dynflow_core.service: Unit not found.
```

So while I feel it should be possible to just start everything, forcing a developer to define what to start is OK too given this response from systemd.

Comment (ekohl) [reply hooks/boot/04-services.rb:20]: > Can this default to ALL_POSSIBLE_SERVICES similar to stop_srevices?

IMHO no since this is mostly used to start services prior to a migration. In that case you only want to start really what's needed.
```
$ rg start_services
spec/services_hook_extensions_spec.rb
15:  context 'start_services' do
16:    subject { context.start_services(services) }

hooks/boot/04-services.rb
20:  def start_services(services)

hooks/pre/30-el7_upgrade_postgresql.rb
2:  start_services(['postgresql'])

katello/hooks/pre/10-reset_feature.rb
8:    start_services(['postgresql'])
```

**Comment by ekohl** [comment]
Updated with corrected indentation in the tests to please Rubocop.

**Review by ehelms — APPROVED** [review]

**Pull request merged**
